File: control/src/main/java/org/openbase/bco/dal/control/layer/unit/agent/AgentControllerFactoryImpl.java
package org.openbase.bco.dal.control.layer.unit.agent;

/*
 * #%L
 * BCO DAL Control
 * %%
 * Copyright (C) 2014 - 2021 openbase.org
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/gpl-3.0.html>.
 * #L%
 */

import org.openbase.bco.dal.lib.layer.unit.agent.Agent;
import org.openbase.bco.dal.lib.layer.unit.agent.AgentController;
import org.openbase.bco.dal.lib.layer.unit.agent.AgentControllerFactory;
import org.openbase.bco.registry.remote.Registries;
import org.openbase.jul.exception.CouldNotPerformException;
import org.openbase.jul.exception.NotAvailableException;
import org.openbase.jul.extension.type.processing.LabelProcessor;
import org.openbase.jul.extension.type.processing.MetaConfigVariableProvider;
import org.openbase.jul.processing.StringProcessor;
import org.openbase.type.domotic.unit.UnitConfigType.UnitConfig;
import org.openbase.type.domotic.unit.agent.AgentClassType.AgentClass;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.lang.reflect.InvocationTargetException;
import java.util.Locale;

/**
 * @author <a href="mailto:divine@openbase.org">Divine Threepwood</a>
 */
public class AgentControllerFactoryImpl implements AgentControllerFactory {

    protected final Logger logger = LoggerFactory.getLogger(AgentControllerFactoryImpl.class);
    private static AgentControllerFactoryImpl instance;

    public static final String META_CONFIG_KEY_AGENT_CLASS_PREFIX = "AGENT_CLASS_PREFIX";

    private static final String PRESET_AGENT_PACKAGE_PREFIX = "org.openbase.bco.app.preset";
    private static final String CUSTOM_AGENT_PACKAGE_PREFIX = "org.openbase.bco.app";

    public synchronized static AgentControllerFactoryImpl getInstance() {
        if (instance == null) {
            instance = new AgentControllerFactoryImpl();
        }
        return instance;
    }

    private AgentControllerFactoryImpl() {
    }

    @Override
    public AgentController newInstance(final UnitConfig agentUnitConfig) throws org.openbase.jul.exception.InstantiationException {
        AgentController agent;
        try {
            if (agentUnitConfig == null) {
                throw new NotAvailableException("AgentConfig");
            }

            Registries.waitForData();
            final AgentClass agentClass = Registries.getClassRegistry().getAgentClassById(agentUnitConfig.getAgentConfig().getAgentClassId());
            final MetaConfigVariableProvider variableProvider = new MetaConfigVariableProvider("AgentClass", agentClass.getMetaConfig());

            final String agentClassPrefix = variableProvider.getValue(
                    META_CONFIG_KEY_AGENT_CLASS_PREFIX,
                    StringProcessor.removeWhiteSpaces(LabelProcessor.getLabelByLanguage(Locale.ENGLISH, agentClass.getLabel())));

            try {
                // try to load preset agent
                String className = PRESET_AGENT_PACKAGE_PREFIX
                        + ".agent"
                        + "." + agentClassPrefix + "Agent";
                agent = (AgentController) Thread.currentThread().getContextClassLoader().loadClass(className).getConstructor().newInstance();
            } catch (ClassNotFoundException | SecurityException | InstantiationException | IllegalAccessException | IllegalArgumentException | NoSuchMethodException | InvocationTargetException ex) {
                // try to load custom agent
                String className = CUSTOM_AGENT_PACKAGE_PREFIX
                        + "." + StringProcessor.removeWhiteSpaces(agentClassPrefix).toLowerCase()
                        + ".agent"
                        + "." + StringProcessor.transformToPascalCase(StringProcessor.removeWhiteSpaces(agentClassPrefix)) + "Agent";
                agent = (AgentController) Thread.currentThread().getContextClassLoader().loadClass(className).getConstructor().newInstance();
            }
            logger.debug("Creating agent of type [" + LabelProcessor.getBestMatch(agentClass.getLabel()) + "]");
            agent.init(agentUnitConfig);
        } catch (CouldNotPerformException | ClassNotFoundException | SecurityException | InstantiationException | IllegalAccessException | IllegalArgumentException | InterruptedException | NoSuchMethodException | InvocationTargetException ex) {
            throw new org.openbase.jul.exception.InstantiationException(Agent.class, agentUnitConfig.getId(), ex);
        }
        return agent;
    }
}

File: module/dal/control/src/main/java/org/openbase/bco/dal/control/layer/unit/agent/AgentControllerFactoryImpl.kt
package org.openbase.bco.dal.control.layer.unit.agent

import org.openbase.bco.dal.lib.layer.unit.agent.Agent
import org.openbase.bco.dal.lib.layer.unit.agent.AgentController
import org.openbase.bco.dal.lib.layer.unit.agent.AgentControllerFactory
import org.openbase.bco.registry.remote.Registries
import org.openbase.jul.exception.CouldNotPerformException
import org.openbase.jul.exception.NotAvailableException
import org.openbase.jul.extension.type.processing.LabelProcessor.getBestMatch
import org.openbase.jul.extension.type.processing.LabelProcessor.getLabelByLanguage
import org.openbase.jul.extension.type.processing.MetaConfigVariableProvider
import org.openbase.jul.processing.StringProcessor
import org.openbase.type.domotic.unit.UnitConfigType
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.lang.reflect.InvocationTargetException
import java.util.*

/**
 * @author [Divine Threepwood](mailto:divine@openbase.org)
 */
class AgentControllerFactoryImpl private constructor() : AgentControllerFactory {
    protected val logger: Logger = LoggerFactory.getLogger(AgentControllerFactoryImpl::class.java)

    @Throws(org.openbase.jul.exception.InstantiationException::class)
    override fun newInstance(agentUnitConfig: UnitConfigType.UnitConfig): AgentController {
        var agent: AgentController
        try {
            if (agentUnitConfig == null) {
                throw NotAvailableException("AgentConfig")
            }

            Registries.waitForData()
            val agentClass = Registries.getClassRegistry().getAgentClassById(agentUnitConfig.agentConfig.agentClassId)
            val variableProvider = MetaConfigVariableProvider("AgentClass", agentClass.metaConfig)

            val agentClassPrefix = variableProvider.getValue(
                META_CONFIG_KEY_AGENT_CLASS_PREFIX,
                StringProcessor.removeWhiteSpaces(getLabelByLanguage(Locale.ENGLISH, agentClass.label))
            )

            try {
                // try to load preset agent
                val className = (PRESET_AGENT_PACKAGE_PREFIX
                        + ".agent"
                        + "." + agentClassPrefix + "Agent")
                agent = Thread.currentThread().contextClassLoader.loadClass(className).getConstructor()
                    .newInstance() as AgentController
            } catch (ex: ClassNotFoundException) {
                // try to load custom agent
                val className = (CUSTOM_AGENT_PACKAGE_PREFIX
                        + "." + StringProcessor.removeWhiteSpaces(agentClassPrefix).lowercase(Locale.getDefault())
                        + ".agent"
                        + "." + StringProcessor.transformToPascalCase(StringProcessor.removeWhiteSpaces(agentClassPrefix)) + "Agent")
                agent = Thread.currentThread().contextClassLoader.loadClass(className).getConstructor()
                    .newInstance() as AgentController
            } catch (ex: SecurityException) {
                val className = (CUSTOM_AGENT_PACKAGE_PREFIX
                        + "." + StringProcessor.removeWhiteSpaces(agentClassPrefix).lowercase(Locale.getDefault())
                        + ".agent"
                        + "." + StringProcessor.transformToPascalCase(StringProcessor.removeWhiteSpaces(agentClassPrefix)) + "Agent")
                agent = Thread.currentThread().contextClassLoader.loadClass(className).getConstructor()
                    .newInstance() as AgentController
            } catch (ex: InstantiationException) {
                val className = (CUSTOM_AGENT_PACKAGE_PREFIX
                        + "." + StringProcessor.removeWhiteSpaces(agentClassPrefix).lowercase(Locale.getDefault())
                        + ".agent"
                        + "." + StringProcessor.transformToPascalCase(StringProcessor.removeWhiteSpaces(agentClassPrefix)) + "Agent")
                agent = Thread.currentThread().contextClassLoader.loadClass(className).getConstructor()
                    .newInstance() as AgentController
            } catch (ex: IllegalAccessException) {
                val className = (CUSTOM_AGENT_PACKAGE_PREFIX
                        + "." + StringProcessor.removeWhiteSpaces(agentClassPrefix).lowercase(Locale.getDefault())
                        + ".agent"
                        + "." + StringProcessor.transformToPascalCase(StringProcessor.removeWhiteSpaces(agentClassPrefix)) + "Agent")
                agent = Thread.currentThread().contextClassLoader.loadClass(className).getConstructor()
                    .newInstance() as AgentController
            } catch (ex: IllegalArgumentException) {
                val className = (CUSTOM_AGENT_PACKAGE_PREFIX
                        + "." + StringProcessor.removeWhiteSpaces(agentClassPrefix).lowercase(Locale.getDefault())
                        + ".agent"
                        + "." + StringProcessor.transformToPascalCase(StringProcessor.removeWhiteSpaces(agentClassPrefix)) + "Agent")
                agent = Thread.currentThread().contextClassLoader.loadClass(className).getConstructor()
                    .newInstance() as AgentController
            } catch (ex: NoSuchMethodException) {
                val className = (CUSTOM_AGENT_PACKAGE_PREFIX
                        + "." + StringProcessor.removeWhiteSpaces(agentClassPrefix).lowercase(Locale.getDefault())
                        + ".agent"
                        + "." + StringProcessor.transformToPascalCase(StringProcessor.removeWhiteSpaces(agentClassPrefix)) + "Agent")
                agent = Thread.currentThread().contextClassLoader.loadClass(className).getConstructor()
                    .newInstance() as AgentController
            } catch (ex: InvocationTargetException) {
                val className = (CUSTOM_AGENT_PACKAGE_PREFIX
                        + "." + StringProcessor.removeWhiteSpaces(agentClassPrefix).lowercase(Locale.getDefault())
                        + ".agent"
                        + "." + StringProcessor.transformToPascalCase(StringProcessor.removeWhiteSpaces(agentClassPrefix)) + "Agent")
                agent = Thread.currentThread().contextClassLoader.loadClass(className).getConstructor()
                    .newInstance() as AgentController
            }
            logger.debug("Creating agent of type [" + getBestMatch(agentClass.label) + "]")
            agent.init(agentUnitConfig)
        } catch (ex: CouldNotPerformException) {
            throw org.openbase.jul.exception.InstantiationException(Agent::class.java, agentUnitConfig.id, ex)
        } catch (ex: ClassNotFoundException) {
            throw org.openbase.jul.exception.InstantiationException(Agent::class.java, agentUnitConfig.id, ex)
        } catch (ex: SecurityException) {
            throw org.openbase.jul.exception.InstantiationException(Agent::class.java, agentUnitConfig.id, ex)
        } catch (ex: InstantiationException) {
            throw org.openbase.jul.exception.InstantiationException(Agent::class.java, agentUnitConfig.id, ex)
        } catch (ex: IllegalAccessException) {
            throw org.openbase.jul.exception.InstantiationException(Agent::class.java, agentUnitConfig.id, ex)
        } catch (ex: IllegalArgumentException) {
            throw org.openbase.jul.exception.InstantiationException(Agent::class.java, agentUnitConfig.id, ex)
        } catch (ex: InterruptedException) {
            throw org.openbase.jul.exception.InstantiationException(Agent::class.java, agentUnitConfig.id, ex)
        } catch (ex: NoSuchMethodException) {
            throw org.openbase.jul.exception.InstantiationException(Agent::class.java, agentUnitConfig.id, ex)
        } catch (ex: InvocationTargetException) {
            throw org.openbase.jul.exception.InstantiationException(Agent::class.java, agentUnitConfig.id, ex)
        }
        return agent
    }

    companion object {
        @get:Synchronized
        var instance: AgentControllerFactoryImpl = AgentControllerFactoryImpl()
            private set

        const val META_CONFIG_KEY_AGENT_CLASS_PREFIX: String = "AGENT_CLASS_PREFIX"

        private const val PRESET_AGENT_PACKAGE_PREFIX = "org.openbase.bco.app.preset"
        private const val CUSTOM_AGENT_PACKAGE_PREFIX = "org.openbase.bco.app"
    }
}
